plc4j/protocols/ads/src/main/java/org/apache/plc4x/java/ads/protocol/Payload2TcpProtocol.java

```diff
@@ -49,21 +50,24 @@ protected void decode(ChannelHandlerContext channelHandlerContext, ByteBuf byteB
             return;
         }
         LOGGER.trace("(-->IN): {}, {}, {}", channelHandlerContext, byteBuf, out);
-        if (byteBuf.readableBytes() < AmsTcpHeader.Reserved.NUM_BYTES + TcpLength.NUM_BYTES) {
-            // wait till we can read the length header
-            return;
+
+        retainingBuf = Unpooled.wrappedBuffer(retainingBuf, byteBuf.retain());
```

Comment:
Why do we need this other Buf? Why not just use the byteBuf directly?
and if its in the case of not being able to read, then netty provides the possibility to "mark" a position and "rollback" to it, see here https://netty.io/4.0/api/io/netty/buffer/ByteBuf.html#markReaderIndex--

Reply:
Every time decode is called it contains a new buffer, all buffers need to be collected until there is sufficient data to parse a message. The buffer needs to be retained the documentation specifies that after the function returns the buffer is released see https://netty.io/4.0/api/io/netty/handler/codec/MessageToMessageDecoder.html . By wrapping the buffers it is possible to keep using it as a buffer. Another option would be to copy the data to a buffer but that sounds less efficient.

Reply:
But the Buffer fills up until its consumed, right? Probably I dont get it right.

Reply:
One can fill a Buffer by writing to it. However in this case you get a new buffer every time, at least I haven't found anything in the documentation about that buffer being filled if not used. It would also contradict the documentation about being able to pass through the buffer if an underlying layer would later modify that same buffer.